src/jit/flowgraph.cpp

```diff
@@ -22472,3 +22476,652 @@ void Compiler::fgLclFldAssign(unsigned lclNum)
+            {
+                GenTreeStmt* finallyRet     = newBlock->lastStmt();
+                GenTreePtr   finallyRetExpr = finallyRet->gtStmtExpr;
+                assert(finallyRetExpr->gtOper == GT_RETFILT);
+                fgRemoveStmt(newBlock, finallyRet);
+                newBlock->bbJumpKind = BBJ_ALWAYS;
+                newBlock->bbJumpDest = normalCallFinallyReturn;
+
+                fgAddRefPred(normalCallFinallyReturn, newBlock);
+            }
+            else
+            {
+                optCopyBlkDest(block, newBlock);
+                optRedirectBlock(newBlock, &blockMap);
+            }
+        }
+
+        // Modify the targeting callfinallies to branch to the cloned
+        // finally. Make a note if we see some calls that can't be
+        // retartged (since they want to return to other places).
+        BasicBlock* const firstCloneBlock    = blockMap[firstBlock];
+        bool              retargetedAllCalls = true;
+
+        for (BasicBlock* block = firstCallFinallyBlock; block != lastCallFinallyBlock; block = block->bbNext)
+        {
+            if (block->isBBCallAlwaysPair())
+            {
+                if (block->bbJumpDest == firstBlock)
+                {
+                    BasicBlock* finallyReturnBlock  = block->bbNext;
+                    BasicBlock* postTryFinallyBlock = finallyReturnBlock->bbJumpDest;
+
+                    if (postTryFinallyBlock == normalCallFinallyReturn)
+                    {
+                        // This call returns to the expected spot, so
+                        // retarget it to branch to the clone.
+                        block->bbJumpDest = firstCloneBlock;
+                        block->bbJumpKind = BBJ_ALWAYS;
+
+                        // For some reason finally entry blocks sometimes
+                        // run a ref count deficit... sigh.
+                        // assert(firstBlock->bbRefs > 0);
+
+                        // fgRemoveRefPred(firstBlock, block);
+                        fgAddRefPred(firstCloneBlock, block);
+                    }
+                    else
+                    {
+                        // We can't retarget this call since it
+                        // returns somewhere else.
+                        retargetedAllCalls = false;
+                    }
+
+                    // Todo -- delete the paired block?
+                }
+            }
+        }
+
+        // If we retargeted all calls, modify EH descriptor to be
+        // try/fault instead of try finally, and then non-cloned
+        // finally catch type to be fault.
+        if (retargetedAllCalls)
+        {
+            JITDUMP("All callfinallies retargeted; changing finally to fault.\n");
+            HBtab->ebdHandlerType  = EH_HANDLER_FAULT;
+            firstBlock->bbCatchTyp = BBCT_FAULT;
+        }
+        else
+        {
+            JITDUMP("Some callfinallies *not* retargeted, so region must remain as a finally.\n");
+        }
```

Comment:
Doesn't this create a new situation in our flow graph where previously all exits from a try/finally would have a call-finally, but now some don't. I'm a little surprised we didn't have an assert about this. Seems like we could end up with illegal flow that we can't detect now that this pattern is legal.